lib/ex_unit/lib/ex_unit/server.ex

```diff
@@ -78,6 +85,7 @@ defmodule ExUnit.Server do
      %{
        state
        | loaded: :done,
+         async_groups: MapSet.new(async_modules, fn {group, _} -> group end),
          async_modules: :queue.from_list(async_modules),
          sync_modules: :queue.from_list(sync_modules)
```

Comment:
I think we can simplify the implementation in the following manner:

1. We will have `async_groups` as a map and `async_modules` as queue
2. Whenever you receive an async module without a group, you store it in the `async_modules` queue as `{:module, case}`
3. Whenever you receive an async module with a group, you check if you know this group, if you don't know it, you put it in the queue and then you store it in the map as `Map.put(async_groups, group, [module])`. If you know the group already, you don't need to put it in the queue, and then you do `Map.update!(async_groups, group, &[module | &1])`
4. Now, `take_async_modules` simply pops from the queue. If it is a module, you send it. If it is a group, you pop it from the map and send it

Reply:
Thanks, that makes sense. I made the changes in https://github.com/elixir-lang/elixir/pull/13897/commits/27b463ddd160f907648613daf6f25448beb42131.